### PR TITLE
support variant, enum, and union derives

### PR DESCRIPTION
**Pull request by dicej**
This is the second stage of implementing #4308.  It adds support for deriving
variant, enum, and union impls for `ComponentType`, `Lift`, and `Lower`.  It
also fixes derived record impls for generic `struct`s, which I had intended to
support in my previous commit, but forgot to test.

Signed-off-by: Joel Dice <joel.dice@fermyon.com>

### CODE REVIEW

File: crates/component-macro/src/lib.rs [hunk 1]
```diff
@@ -230,11 +396,138 @@ impl Expander for LiftExpander {
...
+        let mut lifts = TokenStream::new();
+        let mut loads = TokenStream::new();
+
+        for (index, VariantCase { ident, ty, .. }) in cases.iter().enumerate() {
+            let index_u8 = u8::try_from(index).map_err(|_| {
+                Error::new(
+                    input.ident.span(),
+                    "`enum`s with more than 256 variants not yet supported",
+                )
+            })?;
```
Comment (alexcrichton) [inline crates/component-macro/src/lib.rs:417]: Is this mostly to just simplify this for now where you don't have to worry about 16-vs-8  bit loads?

**Review by alexcrichton — APPROVED** [review]
This looks fantastic to me, thanks for this! I've got one question about the u8 limit, but if it's what I suspect (just more effort needed) mind filing a follow-up issue? I think it's ok to leave as-is for awhile since we're unlikely to get >256 variant-things any time in the near future.

File: crates/component-macro/src/lib.rs [hunk 2]
```diff
@@ -167,9 +190,152 @@ fn expand_record(expander: &dyn Expander, input: DeriveInput) -> Result<TokenStr
...
+                match fields {
+                    syn::Fields::Unnamed(fields) if fields.unnamed.len() == 1 => Ok(VariantCase {
+                        attrs,
+                        ident,
+                        ty: Some(&fields.unnamed[0].ty),
+                    }),
+                    _ => Err(Error::new(
+                        name.span(),
+                        "`union` component types can only be derived for Rust `enum`s \
+                         containing variants with exactly one unnamed field each",
```
[the remark this answers is not in the document]
Comment (dicej) [reply crates/component-macro/src/lib.rs:326]: Did you notice that both `expand_variant` and `expand_union` defer to `Expander::expand_variant`?  So the duplication is already pretty minimal.  But yeah, `expand_variant` and `expand_union` could themselves be combined.

File: crates/component-macro/src/lib.rs (hunk 1 above)
Comment (dicej) [reply crates/component-macro/src/lib.rs:417]: Yeah, I was just keeping it simple to start with.  I was also not sure what the theoretical max was for the component model.  2^32?  More?

File: crates/component-macro/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jameysharp) [reply crates/component-macro/src/lib.rs:326]: I did notice that, although it took me a few read-throughs. :laughing: So yes, I'm suggesting combining `Expander::expand_variant` with `Expander::expand_enum`, except that I know you'll need to take a little care with the implementation for `ComponentType`. But I'm also suggesting combining the top-level `expand_variant`, `expand_union`, and `expand_enum`.

File: crates/component-macro/src/lib.rs (hunk 1 above)
Comment (jameysharp) [reply crates/component-macro/src/lib.rs:417]: The [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#alignment) defines that in the `discriminant_type` function, and the current maximum is 2^32.

File: crates/component-macro/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dicej) [reply crates/component-macro/src/lib.rs:326]: I see -- I'll give that a shot.

[the remark this answers is not in the document]
Comment (dicej) [reply crates/component-macro/src/lib.rs]: Yeah, that slimmed it down a lot.  Great suggestion!

**Review by jameysharp — APPROVED** [review]
I love it. @alexcrichton, want to merge this?

**Pull request merged**